File: rails/x402-micropayments/decode_payment.ts
#!/usr/bin/env node
/**
 * Decode X-PAYMENT header to see what's inside
 *
 * Usage:
 *   npx tsx decode_payment.ts <X-PAYMENT-HEADER-VALUE>
 *
 * Or with ts-node:
 *   npx ts-node decode_payment.ts <X-PAYMENT-HEADER-VALUE>
 *
 * Example:
 *   npx tsx decode_payment.ts "eyJ4NDAyVmVyc2lvbiI..."
 */

const args = process.argv.slice(2);

if (args.length < 1) {
  console.log('Usage: npx tsx decode_payment.ts <X-PAYMENT-HEADER-VALUE>');
  console.log('\nExample:');
  console.log('  npx tsx decode_payment.ts "eyJ4NDAyVmVyc2lvbiI..."');
  process.exit(1);
}

const paymentHeader = args[0];

try {
  // Decode from base64
  const decoded = Buffer.from(paymentHeader, 'base64').toString('utf-8');
  const paymentData = JSON.parse(decoded);

  console.log('=== Decoded X-PAYMENT Header ===\n');
  console.log(JSON.stringify(paymentData, null, 2));

  // Extract key info
  console.log('\n=== Key Information ===');
  if (paymentData.payload?.authorization) {
    const auth = paymentData.payload.authorization;
    console.log(`Payer:         ${auth.payer || 'N/A'}`);
    console.log(`Payee:         ${auth.payee || 'N/A'}`);
    console.log(`Amount:        ${auth.amount || 'N/A'} (atomic units)`);
    console.log(`Nonce:         ${auth.nonce || 'N/A'}`);
    console.log(`Valid Until:   ${auth.validUntil || 'N/A'} (Unix timestamp)`);
    console.log(`Resource:      ${auth.resource || 'N/A'}`);
    console.log(`Signature:     ${auth.signature ? auth.signature.slice(0, 20) + '...' : 'N/A'}`);
  }

  if (paymentData.payload?.requirements) {
    const req = paymentData.payload.requirements;
    console.log(`\nNetwork:       ${req.network || 'N/A'}`);
    console.log(`Asset:         ${req.asset || 'N/A'}`);
    console.log(`Description:   ${req.description || 'N/A'}`);
  }

  console.log('\nNote: The payment header contains both the signed authorization');
  console.log('and the payment requirements for verification by the server.');
} catch (error) {
  console.error('Error decoding payment header:', error);
  process.exit(1);
}

File: rails/x402-micropayments/decode_payment.py
#!/usr/bin/env python3
"""Decode PAYMENT-SIGNATURE header to see what's inside"""

import sys
import base64
import json

if len(sys.argv) < 2:
    print("Usage: python decode_payment.py <PAYMENT-SIGNATURE-HEADER-VALUE>")
    print("\nExample:")
    print('  python decode_payment.py "eyJ4NDAyVmVyc2lvbiI..."')
    sys.exit(1)

payment_header = sys.argv[1]

try:
    decoded = base64.b64decode(payment_header)
    payment_data = json.loads(decoded)

    print("=== Decoded PAYMENT-SIGNATURE Header ===\n")
    print(json.dumps(payment_data, indent=2))

    # Extract key info
    print("\n=== Key Information ===")
    if "payload" in payment_data and "authorization" in payment_data["payload"]:
        auth = payment_data["payload"]["authorization"]
        print(f"Payer:         {auth.get('payer', 'N/A')}")
        print(f"Payee:         {auth.get('payee', 'N/A')}")
        print(f"Amount:        {auth.get('amount', 'N/A')} (atomic units)")
        print(f"Nonce:         {auth.get('nonce', 'N/A')}")
        print(f"Valid Until:   {auth.get('validUntil', 'N/A')} (Unix timestamp)")
        print(f"Resource:      {auth.get('resource', 'N/A')}")
        print(
            f"Signature:     {(auth.get('signature') or '')[:20]}..."
            if auth.get("signature")
            else "Signature:     N/A"
        )

    if "payload" in payment_data and "requirements" in payment_data["payload"]:
        req = payment_data["payload"]["requirements"]
        print(f"\nNetwork:       {req.get('network', 'N/A')}")
        print(f"Asset:         {req.get('asset', 'N/A')}")
        print(f"Description:   {req.get('description', 'N/A')}")

    print("\nNote: The payment header contains both the signed authorization")
    print("and the payment requirements for verification by the server.")

except Exception as e:
    print(f"Error decoding payment header: {e}")
    sys.exit(1)
